src/ShaderGen.Tests/ShaderBuiltinsTests.cs

```diff
@@ -0,0 +1,476 @@
+        {
+            if (!toolChain.Features.HasFlag(ToolFeatures.ToHeadless))
+                throw new RequiredToolFeatureMissingException(
+                    $"The {toolChain} does not support creating a headless graphics device!");
+
+            string csFunctionName =
+                $"{nameof(TestShaders)}.{nameof(ShaderBuiltinsComputeTest)}.{nameof(ShaderBuiltinsComputeTest.CS)}";
+            Compilation compilation = TestUtil.GetCompilation();
+
+            LanguageBackend backend = toolChain.CreateBackend(compilation);
+
+            /*
+             * Compile backend
+             */
+            ShaderSetProcessor processor = new ShaderSetProcessor();
+
+            ShaderGenerator sg = new ShaderGenerator(
+                compilation,
+                backend,
+                null,
+                null,
+                csFunctionName,
+                processor);
+
+            ShaderGenerationResult generationResult = sg.GenerateShaders();
+            GeneratedShaderSet set = generationResult.GetOutput(backend).Single();
+            _output.WriteLine($"Generated shader set for {toolChain.Name} backend.");
+
+            CompileResult compilationResult =
+                toolChain.Compile(set.ComputeShaderCode, Stage.Compute, set.ComputeFunction.Name);
+            if (compilationResult.HasError)
+            {
+                _output.WriteLine($"Failed to compile Compute Shader from set \"{set.Name}\"!");
+                _output.WriteLine(compilationResult.ToString());
+                Assert.True(false);
+            }
+            else
+                _output.WriteLine($"Compiled Compute Shader from set \"{set.Name}\"!");
+
+            Assert.NotNull(compilationResult.CompiledOutput);
+
+            int sizeOfParametersStruct = Unsafe.SizeOf<ComputeShaderParameters>();
+
+            // Create failure data structure, first by method #, then by field name.
+            Dictionary<int, List<Tuple<ComputeShaderParameters, ComputeShaderParameters,
+                IReadOnlyCollection<Tuple<string, float, float>>>>> failures
+                = new Dictionary<int, List<Tuple<ComputeShaderParameters, ComputeShaderParameters,
+                    IReadOnlyCollection<Tuple<string, float, float>>>>>();
+
+            // We need two copies, one for the CPU & one for GPU
+            ComputeShaderParameters[] cpuParameters = new ComputeShaderParameters[ShaderBuiltinsComputeTest.Methods];
+            ComputeShaderParameters[] gpuParameters = new ComputeShaderParameters[ShaderBuiltinsComputeTest.Methods];
+            int loops = 0;
+            long durationTicks;
+
+            // Set start.
+            long startTicks = Stopwatch.GetTimestamp();
+
+            ShaderBuiltinsComputeTest cpuTest = new ShaderBuiltinsComputeTest();
+            /*
+             * Run shader on GPU.
+             */
+            using (GraphicsDevice graphicsDevice = toolChain.CreateHeadless())
```

Comment:
I hope this doesn't come across badly, but I really don't like the `DisposeCollectorResourceFactory` because it doesn't implement `IDisposable` itself, which prevents it being used in a `using` block.  Instead you have to create it and manually call `DisposeCollectorResourceFactory.DisposeCollector.DisposeAll` and wrap in a `try...finally` to be safe and gaurantee disposal is always called.  This is particularly important as, I humbly submit, many of the `Veldrid` classes don't fully implement the disposal pattern 'properly' (at least as recommended by Microsoft for lots of good reasons), particularly when using unmanaged resources, which can cause all kinds of clean up issues, memory leaks, etc. and are very hard bugs to track down.  It only takes a minute to find examples, e.g. [D3D11Fence](https://github.com/mellinoe/veldrid/blob/ece7f27f175c130c825ef28e7bd4b7237831c94f/src/Veldrid/D3D11/D3D11Fence.cs#L22-L25).

For an excellent explanation of how this, and similar classes, should be implement `IDisposable` consider [this Microsoft article](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose).  Most of the systems I write are designed for up times measured in years so I've always been a bit overly anal about proper disposal pattern implementation.  As `Veldrid` makes use of so many unmanaged resources, I'd recommend a review of the use of those resources, and proper implementation of the disposal pattern where appropriate, at some point.

One trick I prefer, that the article doesn't cover is using code similar to:
```csharp
Interlocked.Exchange(ref _myDisposableField, null)?.Dispose();
```
-as a simple, compact, and thread-safe way to prevent double-disposal.  Another consideration, particularly for game engines, is the importance of realising the performance impact of adding finalisers.  The solution is to avoid disposing objects and instead using them in a pool where appropriate (this is good for GC anyway).


As for reducing nesting, the six usings are deliberately placed seqentially to avoid the necessity for nesting blocks, so actually using the  `DisposeCollectorResourceFactory`  wouldn't decrease the nesting visually anyway.

However, if you are insistent, I will change the code to use it, just let me know!

Reply:
> This is particularly important as, I humbly submit, many of the Veldrid classes don't fully implement the disposal pattern 'properly' (at least as recommended by Microsoft for lots of good reasons), particularly when using unmanaged resources, which can cause all kinds of clean up issues, memory leaks, etc. and are very hard bugs to track down. It only takes a minute to find examples, e.g. D3D11Fence.

That guideline is pretty archaic, and frankly doesn't work for Veldrid. Finalizers are not deterministic, but the order of disposal in Veldrid cannot be arbitrary (child objects need to be disposed before their parents). I think this is a common problem with disposable resources, not just the ones in graphics API's, which has always made me very skeptical of that pattern. It's been a conscious decision to avoid it in Veldrid.

> As for reducing nesting, the six usings are deliberately placed seqentially to avoid the necessity for nesting blocks, so actually using the DisposeCollectorResourceFactory wouldn't decrease the nesting visually anyway.

I was worried about VS insisting on indenting the nested usings, but it's okay with them as-is. It's still a bit hard to determine the scope of the usings, but I won't insist on anything.